proposals/BVAL-214.md

```diff
@@ -0,0 +1,197 @@
+
+The values will be exposed through a generic property retrieval API:
+
+    @PasswordsMatch
+    public class UserDataBean {
+        String name;
+        String password;
+        String passwordRepeat;
+    }
+
+    public class PasswordMatchValidator implements ConstraintValidator<PasswordsMatch, UserDataBean> {
+
+        void initialize(PasswordsMatch annotation) {
+        }
+
+        boolean isValid(UserDataBean value, ConstraintValidatorContext ctx) {
+            String password = null;
+            String passwordRepeat = null;
+
+            // value null can have two reasons here:
+            // a) a null reference / collection entry during cascaded validation
+            // b) validateValues(); In this case we can get the property values from the context
+            if ( value == null ) {
+                password = (String) ctx.property( "password" ).get();
+                passwordRepeat = (String) ctx.property( "passwordRepeat" ).get();
+            }
+            else {
+                password = value.password;
+                passwordRepeat = value.passwordRepeat;
+            }
+
+            // TODO Make null-safe
+            return password.equals( passwordRepeat );
+        }
+    }
+
+The values would be exposed through the constraint validator context. The API would mirror the one using for passing values:
+
+    ctx.property( "name" ).get();
+    ctx.list( "emails" ).index( 1 ).get();
+    ctx.map( "addresses" ).key( "home" ).get();
+    ctx.property( "phone" ).property( "areaCode" ).get();
+    ctx.map( "addresses" ).key( "home" ).property( "street" ).get();
+
+**Pro:**
+
+* No requirement for clonability towards validated bean types
+* No bean instance needed, resembles more closely the current `validateValue()` method
+* Enables much simpler cross-field constraints (see below)
+
+**Cons:**
+
+* Solution is not transparent to class-level constraint validators, they must account for the fact that values are to be obtained through the context; I think it's ok, but existing validators need updating.
+
+This proposal enables cross-field constraints nicely:
+
+    public class UserDataBean {
+        String name;
+        String password;
+
+        @Equals("password")
+        String passwordRepeat;
+    }
+
+    public class EqualsValidator implements ConstraintValidator<Equals, String> {
+
+        private String compareTo;
+
+        void initialize(Equals annotation) {
+            this.compareTo = annotation.value();
+        }
+
+        boolean isValid(String value, ConstraintValidatorContext ctx) {
+            if  ( value == null ) {
+                return true;
+            }
+
+            String comparedValue = (String) ctx.property( compareTo ).get();
+            return value.equals( comparedValue );
+        }
+    }
+
+That's nicer than the traditional class-level constraint. The good thing is that it'd work automatically in both cases:
```

Comment:
Are you basically saying that the context always carries all properties, even for `validate()`? I guess that would be a requirement to have some sort of consistency, right? Or can you only access the properties in the case of class level constraints? 

Is the retrieval of the properties dynamic? Would the context retrieve the properties on request from the bean instance (except the case of validateValues)?